### PR TITLE
Add testnet stake pool to stake-o-matic.sh

### PR DESCRIPTION
**Pull request by joncinque**
#### Problem

In order to have stake pools run nicely next to not-a-stake-pool, let's reuse a lot of the plumbing.

#### Solution

Add `testnet-stake-pool` as a cluster option for use with buildkite.  Even though the variable `RESERVE_ACCOUNT_ADDRESS` is poorly named for stake pools, since it actually represents the address of the pool, it seems like it's best to maintain the same name to work with the current setup.

We'll just need to add another step to run the testnet-stake-pool version of stake-o-matic to buildkite.

### CODE REVIEW

File: bot/src/main.rs [hunk 1]
```diff
@@ -598,7 +598,7 @@ fn get_config() -> BoxResult<(Config, RpcClient, Box<dyn GenericStakePool>)> {
                     .long("baseline-stake-amount")
                     .value_name("SOL")
                     .takes_value(true)
-                    .default_value("5000")
+                    .default_value("5")
```
Comment (joncinque) [inline bot/src/main.rs:601]: This is a safe default amount while we're working with small stakes

File: stake-o-matic.sh [hunk 2]
```diff
@@ -52,18 +48,32 @@ MAINNET_BETA_ARGS=(
...
   ${STAKE_AUTHORITY_KEYPAIR:?}
 )
 
-if [[ $CLUSTER == "testnet" ]]; then
-  ./solana-stake-o-matic "${TESTNET_ARGS[@]}"
+# shellcheck disable=SC2206
+STAKE_POOL_ARGS=(
+  stake-pool
+  --baseline-stake-amount ${BASELINE_STAKE_AMOUNT:?}
+  ${RESERVE_ACCOUNT_ADDRESS:?}
```
Comment (joncinque) [inline stake-o-matic.sh:65]: This variable is poorly named and should really be `${STAKE_POOL_ADDRESS:?}`, but it seems like that wouldn't place with the dynamic setup

**Comment by danpaul000** [comment]
Will running `testnet-stake-pool` configuration break anything with the "testnet stake requirement" that @mvines built for the mainnet stake criteria?

File: stake-o-matic.sh (hunk 2 above)
Comment (danpaul000) [reply stake-o-matic.sh:65]: You can make the renaming, and just add a fourth cluster config `mainnet-beta-stake-pool` that uses it, and `mainnet-beta` can continue to use `stake-pool-v0` args

File: bot/src/main.rs (hunk 1 above)
Comment (mvines) [reply bot/src/main.rs:601]: Should we just take out the default value entirely?

**Review by mvines — COMMENTED** [review]
> Will running `testnet-stake-pool` configuration break anything with the "testnet stake requirement" that @mvines built for the mainnet stake criteria?

We could use a different _db_ location for `testnet-stake-pool`:
https://github.com/solana-labs/stake-o-matic/blob/13e79e8527c7a4aefa8aeb057a2fe7c54ab129f1/stake-o-matic.sh#L16
as writing the `testnet-stake-pool` markdown files in the current db but at a different path will still cause them to show up in the wiki and confuse the existing users.  Maybe call it `solana-labs/testnet-stake-pool-o-matic`

File: bot/src/main.rs (hunk 1 above)
Comment (joncinque) [reply bot/src/main.rs:601]: Yeah, it's always being set anyway

**Comment by danpaul000** [comment]
> > Will running `testnet-stake-pool` configuration break anything with the "testnet stake requirement" that @mvines built for the mainnet stake criteria?
> 
> We could use a different _db_ location for `testnet-stake-pool`:
> 
> https://github.com/solana-labs/stake-o-matic/blob/13e79e8527c7a4aefa8aeb057a2fe7c54ab129f1/stake-o-matic.sh#L16
> 
> as writing the `testnet-stake-pool` markdown files in the current db but at a different path will still cause them to show up in the wiki and confuse the existing users. Maybe call it `solana-labs/testnet-stake-pool-o-matic`

the idea being the stake-pool-v0 will continue to run on testnet for a while acting as the pre-req for MB stake checks, and the stake-pool will run on testnet in parallel with a different pile of tokens?

**Comment by joncinque** [comment]
> Maybe call it solana-labs/testnet-stake-pool-o-matic

@mvines Just so I understand, do you mean updating the `cluster_db_path_for()` function to also to take in some other parameter (stake-o-matic type?), and then it can write out normally for not-a-stake-pool, and write to cluster + `stake-pool-o-matic`for stake-pool?  Or do you want to have the stake pool wiki in a totally separate repo? It seems like it would be ok to host it in a different subdirectory within this wiki's repo, but I'll defer to you on this.

**Comment by mvines** [comment]
Different subdirectory in this wiki works as long as we don’t want add the —markdown flag (which may make monitoring harder but up to you).  With the —markdown flag, we’ll get duplicate entries in the wiki for each testnet validator and the Testnet index page displayed in the wiki output that GitHub renders

**Comment by joncinque** [comment]
> With the —markdown flag, we’ll get duplicate entries in the wiki for each testnet validator and the Testnet index page displayed in the wiki output that GitHub renders

Good point, I'll add the same logic for the markdown path

**Comment by mvines** [comment]
A different path for the markdown won’t help.  The paths are flattened when rendered in the GitHub wiki UI.  So it’s either separate wiki project, or no markdown for the stake-pool test pools

**Comment by joncinque** [comment]
> A different path for the markdown won’t help. The paths are flattened when rendered in the GitHub wiki UI. So it’s either separate wiki project, or no markdown for the stake-pool test pools

I'm just getting every part of this wrong :sweat_smile: as an alternative, I was thinking we could create a custom sidebar with groups, but it seems like it's more trouble than it's worth.  I'll just disable markdown generation.  Adding some sort of prefix to separate not-a-stake-pool and stake-pool files would probably cause more confusion

**Comment by joncinque** [comment]
This takes a bit of a different turn to support not-a-stake-pool :tm: and stake-pool :tm: at the same time.  The main idea is to have stake pools running regularly on buildkite without messing up the wiki, but also allowing me to inspect the results easily.  To that end, this adds a noop mode to simply take a database file, generate markdown, and allow me to see how it's running in a more friendly manner offline.  Here's what it requires:

* noop stake pool: this is a special mode for generating markdown from an already existing epoch db
* markdown mode: instead of only generating markdown on the first run, it's possible to specify "yes" (always), "no" (never), and "first" (only on first time)
* db suffix: allows the normal "data-testnet" to coexist with "data-testnet-stake-pool"

Otherwise, a few other bits:
* make baseline amount mandatory
* rename the shell variable for stake pools: `RESERVE_STAKE_ADDRESS` -> `STAKE_POOL_ADDRESS`
* add dry run support to the rest of stake pools

**Review by mvines — APPROVED** [review]
lgtm!

**Pull request merged**